File: src/121. Best Time to Buy and Sell Stock/121.js
/*
 * @Author: Nick Cao
 * @Date: 2021-06-06 11:48:06
 * @LastEditTime: 2021-06-06 11:59:17
 * @LastEditors: Nick Cao
 * @Description: 
 * @FilePath: \LeetCodeLog\src\121. Best Time to Buy and Sell Stock\121.js
 */


/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
    let lowest = Infinity;
    let profit = 0;
    for (const price of prices) {
        // update buy timing
        if (price < lowest) {
            lowest = price;
        }
        // update profit
        profit = Math.max(profit, price - lowest);
    }
    return profit;
};

File: src/121.BestTimetoBuyandSellStock/121.rs
/*
 * @Author: Nick Cao
 * @Date: 2021-06-06 14:18:45
 * @LastEditTime: 2021-06-06 14:25:19
 * @LastEditors: Nick Cao
 * @Description: 
 * @FilePath: \LeetCodeLog\src\121. Best Time to Buy and Sell Stock\121.rs
 */

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut buy = prices[0];
        let mut profit = 0;
        for price in prices.iter() {
            if buy > *price {
                buy = *price;
            } else if profit < *price - buy {
                profit = *price - buy;
            }
        }
        profit
    }
}